src/runtime/program/bpf_loader_program/execute.zig

```diff
@@ -0,0 +1,2506 @@
+    defer programdata.release();
+
+    try programdata.serializeIntoAccountData(bpf_loader_program.v3.State{
+        .program_data = .{
+            .slot = clock.slot,
+            .upgrade_authority_address = upgrade_authority_address,
+        },
+    });
+
+    try ic.tc.log("Extended ProgramData account by {} bytes", .{additional_bytes});
+}
+
+/// [agave] https://github.com/anza-xyz/agave/blob/01e50dc39bde9a37a9f15d64069459fe7502ec3e/programs/bpf_loader/src/lib.rs#L1346-L1515
+pub fn executeV3Migrate(
+    allocator: std.mem.Allocator,
+    ic: *InstructionContext,
+) !void {
+    if (!ic.tc.feature_set.active.contains(
+        feature_set.ENABLE_LOADER_V4,
+    )) {
+        return InstructionError.InvalidInstructionData;
+    }
+
+    const AccountIndex = bpf_loader_program.v3.instruction.Migrate.AccountIndex;
+    try ic.info.checkNumberOfAccounts(3);
+
+    const programdata_key = ic.getAccountKeyByIndex(@intFromEnum(AccountIndex.program_data));
+    const program_key = ic.getAccountKeyByIndex(@intFromEnum(AccountIndex.program));
+    const provided_authority_key = ic.getAccountKeyByIndex(@intFromEnum(AccountIndex.authority));
+
+    // [agave] https://github.com/anza-xyz/agave/blob/5fa721b3b27c7ba33e5b0e1c55326241bb403bb1/program-runtime/src/sysvar_cache.rs#L130-L141
+    const clock = ic.tc.sysvar_cache.get(sysvar.Clock) orelse
+        return InstructionError.UnsupportedSysvar;
+
+    // Verify ProgramData account.
+    const progdata_info = info: {
+        var programdata = try ic.borrowInstructionAccount(@intFromEnum(AccountIndex.program_data));
+        defer programdata.release();
+
+        if (!programdata.context.is_writable) {
+            try ic.tc.log("ProgramData account not writeable", .{});
+            return InstructionError.InvalidArgument;
+        }
+
+        const program_len, const upgrade_key = switch (try programdata.deserializeFromAccountData(
+            allocator,
+            bpf_loader_program.v3.State,
+        )) {
+            .program_data => |data| blk: {
+                if (clock.slot == data.slot) {
+                    try ic.tc.log("Program was deployed in this block already", .{});
+                    return InstructionError.InvalidArgument;
+                }
+
+                const program_len = programdata.constAccountData().len -|
+                    bpf_loader_program.v3.State.PROGRAM_DATA_METADATA_SIZE;
+                break :blk .{ program_len, data.upgrade_authority_address };
+            },
+            else => .{ 0, null },
+        };
+
+        break :info .{
+            .len = program_len,
+            .upgrade_key = upgrade_key,
+            .funds = programdata.account.lamports,
+        };
+    };
+
+    // [agave] https://github.com/anza-xyz/agave/blob/01e50dc39bde9a37a9f15d64069459fe7502ec3e/programs/bpf_loader/src/lib.rs#L399-L401
+    const migration_authority =
+        Pubkey.parseBase58String("3Scf35jMNk2xXBD6areNjgMtXgp5ZspDhms8vdcbzC42") catch unreachable;
```

Comment:
since this is a const related to the whole program - should we keep it at the top of the file? - i understand its nice to have close to its actual usage but the inconsistency to how we handle all other global consts in the codebase seems a bit off - wdyt

Reply:
https://github.com/Syndica/sig/pull/609/commits/8c0e97d352bd3e781a25609e2c5033e0bfc02efb